util/redis.js

```diff
@@ -16,6 +16,17 @@ exports = module.exports = {
       host: config.redis_host,
       port: config.redis_port,
       password: config.redis_password,
+      reconnectOnError(err) {
+        /* Reconnect when ElastiCache has promoted some other node to primary & 
+         * demoted the node we are connected to a replica.
+         * https://github.com/redis/ioredis/blob/main/README.md#reconnect-on-error
+         */
+        var targetError = 'READONLY';
+        if (err.message.slice(0, targetError.length) === targetError) {
+          config.logger.info(`Reconnecting due to error: ${err.message}`);
+          return true;
```

Comment:
I wonder if this should be `2` to [resend the failed command](https://redis.github.io/ioredis/interfaces/CommonRedisOptions.html#reconnectOnError)? Otherwise... it reconnects but the command fails? Then we're just depending on eos-phone-home to send another request?

I 100% agree about the integration test. What I was thinking about doing was moving the redis container setup into the tests as a fixture. I know how I'd do that with pytest. No clue with chai. Then again, it seems you have to have a cluster to make the replicas read only. Seems like a pain.

Reply:
> I wonder if this should be `2` to [resend the failed command](https://redis.github.io/ioredis/interfaces/CommonRedisOptions.html#reconnectOnError)? Otherwise... it reconnects but the command fails? Then we're just depending on eos-phone-home to send another request?

You're probably right: retrying in the service would be better. eos-phone-home will try again but not for at least 3 hours.

> I 100% agree about the integration test. What I was thinking about doing was moving the redis container setup into the tests as a fixture. I know how I'd do that with pytest. No clue with chai. Then again, it seems you have to have a cluster to make the replicas read only. Seems like a pain.

I learned that if you have two redises on ports 6379 and 6380 for instance, you can make one a replica of the other with:

```
redis-cli -p 6379 replicaof localhost 6380
```

And you can swap them with something like:

```
redis-cli -p 6379 replicaof no one
redis-cli -p 6380 replicaof localhost 6379
```

But the issue is that our service will always try to reconnect to the same (host, port). Unlike in the AWS case where they change the `master.$FOO` subdomain to point to the current primary.

For testing I would just swap them back after a while so it would eventually reconnect to the newly reprimarified primary.

Reply:
Yeah, I figured. [This page](https://www.dragonflydb.io/faq/setup-read-only-replication-redis) suggests that you can dynamically toggle the replicas as read only with `CONFIG SET replica-read-only yes/no` (that page uses the older slave naming). So, you could start the replica on 6379, change it to not read only, start eos-activation-server, change it to be read only, and then make a request to your new `/health` endpoint.

Reply:
That works. Here's the result of hitting the `/health` endpoint after configuring the Redis on 6379 to be a read-only replica:

```
2024-09-30T10:32:46.121+00:00 - info: Reconnecting due to error: READONLY You can't write against a read only replica.
2024-09-30T10:32:46.125+00:00 - info: Disconnected from Redis server
2024-09-30T10:32:46.126+00:00 - info: Reconnecting to 127.0.0.1:6379 in 50 ms
2024-09-30T10:32:46.178+00:00 - info: Connected to Redis server
2024-09-30T10:32:46.179+00:00 - info: Redis connection ready
```

and the response (I'm running the server with `NODE_ENV=test`):

```
HTTP/1.1 500 Internal Server Error
Connection: keep-alive
Content-Length: 98
Content-Type: application/json; charset=utf-8
Date: Mon, 30 Sep 2024 10:33:41 GMT
ETag: W/"62-clhYtO7g5QlpvP1M7BtWjK/1zjQ"
Keep-Alive: timeout=5
```

```json
{
    "error": "ReplyError: READONLY You can't write against a read only replica.",
    "status": "unhealthy"
}
```

If I make the replica read-write, the `/health` endpoint is healthful.

Reply:
If I return `2` to resend the command, then until I flip the replica back to read-write the activation server reconnects in a tight loop. There is an exponential backoff for reconnecting when the initial connection fails, but not in this case of application-level: the delay is always 50 ms.

So to me that seems unwise.

Reply:
I left it returning true; but added a comment as to what returning 2 would do and why I've not done so.

Reply:
https://node.testcontainers.org/quickstart/ looks like a potentially good way to move the handling of the redis instance into the test.